backend/src/ml_space_lambda/app_configuration/lambda_functions.py

```diff
@@ -18,39 +18,234 @@
+            version_id=version_id,
+            configuration=new_configuration,
+            changed_by=event["requestContext"]["authorizer"]["principalId"],
+            change_reason=request["changeReason"],
+            created_at=time.time(),
+        )
+
+        try:
+            # Create a record for this configuration and mark it the latest config by incrementing the version Id
+            app_configuration_dao.create(config=app_configuration)
+        except Exception as e:
+            if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
+                e.response["ResponseMetadata"]["HTTPStatusCode"] = 429
+            raise e
 
-    return f"Successfully updated configuration for {configScope}, version {version_id}."
+        # Create the new app deny policy
+        try:
+            deny_policy_statements = []
+            enabled_services_dict = app_configuration.configuration.enabled_services.to_dict()
+            # Check each service
+            for service in enabled_services_dict.keys():
+                # If the service has permissions that need to be denied
+                if service in service_disable_permissions:
+                    # If the service is deactivated
+                    if not enabled_services_dict[service]:
+                        if "ResourceType" in service_disable_permissions[service]:
+                            suspend_all_of_type(service_disable_permissions[service]["ResourceType"])
+                        if "Statements" in service_disable_permissions[service]:
+                            deny_policy_statements.extend(service_disable_permissions[service]["Statements"])
+
+            # Check for permissions associated with multiple services
+            for group in service_group_disable_permissions:
+                # If all the services in a group are deactivated, then add the group deny statements to the deny policy
+                if "ServiceGroup" in group:
+                    active_service_in_group = False
+                    for group_service in group["ServiceGroup"]:
+                        if enabled_services_dict[group_service]:
+                            active_service_in_group = True
+                            break
+                    # Deny permissions related to the service group
+                    if not active_service_in_group and "Statements" in group:
+                        deny_policy_statements.extend(group["Statements"])
+
+            # When all services are activated, apply a filler statement so that the policy does not need to be unattached and deleted
+            # Unattaching from all roles could cause throttling and will require re-attaching to all dynamic roles later
+            if len(deny_policy_statements) == 0:
+                deny_policy_statements.extend(filler_deny_statement)
+
+            # Will create, update the current deny services policy with the new permissions
+            iam_manager.update_dynamic_policy(
```

Comment:
I am not super familiar with how this works but what happens if the users bring in their own IAM policies?

Reply:
Great call! This should all be in a conditional for if they are using dynamic roles. If they are using dynamic roles, then this should always work as expected.